SignalMessaging/utils/OWSGroupAvatarBuilder.m

```diff
@@ -32,19 +38,51 @@ - (nullable UIImage *)buildSavedImage
     return self.thread.groupModel.groupImage;
 }
 
++ (OWSContactsManager *)contactsManager
+{
+    return (OWSContactsManager *)SSKEnvironment.shared.contactsManager;
+}
+
 - (nullable UIImage *)buildDefaultImage
 {
-    return self.class.defaultGroupImage;
+    return [self.class defaultAvatarForGroupId:self.thread.groupModel.groupId
+                         conversationColorName:self.thread.conversationColorName
+                                      diameter:self.diameter];
+}
+
++ (nullable UIImage *)defaultAvatarForGroupId:(NSData *)groupId
+                        conversationColorName:(NSString *)conversationColorName
+                                     diameter:(NSUInteger)diameter
+{
+    NSString *cacheKey = [NSString stringWithFormat:@"%@-%d", groupId.hexadecimalString, Theme.isDarkThemeEnabled];
+
+    UIImage *_Nullable cachedAvatar =
+        [OWSGroupAvatarBuilder.contactsManager.avatarCache imageForKey:cacheKey diameter:(CGFloat)diameter];
+    if (cachedAvatar) {
+        return cachedAvatar;
+    }
+
+    UIColor *backgroundColor =
+        [UIColor ows_conversationColorForColorName:conversationColorName isShaded:Theme.isDarkThemeEnabled];
+    UIImage *_Nullable image =
+        [OWSGroupAvatarBuilder groupAvatarImageWithBackgroundColor:backgroundColor diameter:diameter];
+    if (!image) {
+        return nil;
+    }
+
+    [OWSGroupAvatarBuilder.contactsManager.avatarCache setImage:image forKey:cacheKey diameter:diameter];
+    return image;
 }
 
-+ (UIImage *)defaultGroupImage
++ (nullable UIImage *)groupAvatarImageWithBackgroundColor:(UIColor *)backgroundColor diameter:(NSUInteger)diameter
 {
-    static UIImage *defaultGroupImage;
-    static dispatch_once_t onceToken;
-    dispatch_once(&onceToken, ^{
-        defaultGroupImage = [UIImage imageNamed:@"empty-group-avatar"];
-    });
-    return defaultGroupImage;
+    UIImage *icon = [UIImage imageNamed:@"group-avatar"];
```

Comment:
Hopefully it looks good at the variety of dimensions we use, otherwise we should consider having multiple assets that have been sized/hinted per resolution.

Reply:
I think it looks reasonable. WDYT?

![screen shot 2018-09-27 at 9 06 09 am](https://user-images.githubusercontent.com/36971147/46147856-a2c97600-c234-11e8-9e13-9c1f8d08ca2d.png)